### DISCUSSION TITLE
STM32F401 SPI2 and SPI1 at the same time

### DISCUSSION THREAD

**Discussion by fpgiovanini** [stm32duino/Arduino_Core_STM32#1626] [Libraries]
I need to use two SPI ports in a STM32F401. (SPI1 (PA7, PA6, PA5)😉 and SPI2 (PB15, PB14, PB13)  MOSI  MISO  SCLK
pins repectively.

I declared SPIClass SPI_2(PB15, PB14, PB13); in my sketch and it works, but SPI1 don´t work and craches the processor.

SPI.begin();
SPI_2.begin(); //Initiallize the SPI 2
SPI_2.beginTransaction(SPI_SETTINGS)

**Comment by ABOSTM** [comment]
> Hi Alexander.
> I´m glad to hear from you.Thanks.
> 
> Did you try to use SPI2 only ? A: yes
> does it work ?A: yes
> 
> 
> Also can you check each of your SPI device on SPI1 in stand alone (without other SPI device),
> to make sure this is not the device which causes issue?
> A: the device works well on SPI1 of other uProcessor.
> I atacched the schematic to you have an idea of the project.
> ESP32 two SPI´s ports --work well
> STM32F401 only SPI2 works -SPI1 CRACHES THE PROCESSOR.
> 
> Thanks a lot, best regards!!
> 
> fernando

Can you also try to use SPI1 only ? does it work ?
My question remain valid (I changed SPI instance): 

> Also can you check each of your SPI device on **SPI2** in stand alone (without other SPI device),

Point is not to check whether SPI device is well working but especially if it works with STM32.
The aims of those question is to try to figure out what is wrong: is it SPI1? is it the fact to have 2 SPI in parallel? or is it linked to STM32 with a specific SPI Device ?
Note I have not found schematics you talk about.

**Comment by fpgiovanini** [comment]
Hi Alexander.
[lt1799.pdf](https://github.com/stm32duino/Arduino_Core_STM32/files/7882375/lt1799.pdf)
Please, take a look in the schematic.
Thanks.

**Reply by ABOSTM** [reply]
Hi @fpgiovanini, any news about SPI1 only tests ?

>  Can you also try to use SPI1 only ? does it work ?

**Comment by fpgiovanini** [comment]
Hi Mr. Bourdiol

When I only use  SPI1, it OK .
The problem is to use SPI1 and SPI2 at the same time.

Thanks  a lot, best regards

fernando





Enviado do Outlook<http://aka.ms/weboutlook>

________________________________
De: Alexandre Bourdiol ***@***.***>
Enviado: segunda-feira, 24 de janeiro de 2022 05:53
Para: stm32duino/Arduino_Core_STM32 ***@***.***>
Cc: fpgiovanini ***@***.***>; Mention ***@***.***>
Assunto: Re: [stm32duino/Arduino_Core_STM32] STM32F401 SPI2 and SPI1 at the same time (Discussion #1626)


Hi @fpgiovanini<https://github.com/fpgiovanini>, any news about SPI1 only tests ?

Can you also try to use SPI1 only ? does it work ?

—
Reply to this email directly, view it on GitHub<https://github.com/stm32duino/Arduino_Core_STM32/discussions/1626#discussioncomment-2031863>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ALS7Q7SWVIRFC2ZQVX6BXP3UXUHPJANCNFSM5METBLAA>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.

**Comment by ABOSTM** [comment]
Hi @fpgiovanini,
I made a sketch on my nucleo-F401RE with both SPI1 (PA7, PA6, PA5) and SPI2 (PB15, PB14, PB13),
and both work fine, alternating in loop SPI1 (graphic LCD) and SPI2  (SD card)
Note that In addition I used a different SS pin D10 for SPI1 and PC8 for SPI2

So I guess there is something wrong, either in your hardware or in your software.
Can you verify your hardware ?
Can you share your sketch (minimal one with only SPI1 and SPI2, which reproduce the issue) ?

**Reply by lockie555** [reply]
Hi@ABOSTM
Can you show the sketch? I have tried code examples from a number of sources but all have failed to compile on Arduino.IDE.
I am using nucleo-L152 and I want to use SPI2 concurrently with SPI1, but I cannot seem to get SPI2 to work with any code - always getting errors like    xxx not declared n this scope. Everything works fine using SPI1 only, or SPI1 and I2C, but not SPI and SPI2 together.
I would like to try your code to see if I have a hardware issue.
Thanks

**Reply by ABOSTM** [reply]
Hi @lockie555,
Sorry, I don't have my source code any more, and it would probably not help as it is a different Hardware MCU, and different issue.

Did you try to use SPI2 solely ?
As it seems you have a compilation error, please provide your sketch (minimal sketch reproducing the issue) and the build error message.
By the way I propose you open a new issue or discussion, as I think it is different from the current one.

**Comment by fpgiovanini** [comment]
[STM32F401_SPI.pdf](https://github.com/stm32duino/Arduino_Core_STM32/files/7928884/STM32F401_SPI.pdf)
Hi Alexandre.
I´m glad to hear you were successful using SPI1 and SPI2 simultaneously in the F401 nucleo.
I think the hardware is ok, because is the same in the esp32 board and works fine.
I attached me minimal sketch I have been testing unsuccessfully.
If you find and mistake, please send me the corrections.
Thanks a lot for your help, best regards.

**Comment by fpgiovanini** [comment]
Hi Alexandre
I think I found the reason of strange behavor of F401.
Is it possible that the parameter SPI_CONTINUE causes all the problems in the subroutine?
I would like to apologize for my stupidity  and for you wasting your valuable time.
Best regards, thanks a lot.

**Comment by ABOSTM** [comment]
Hi @fpgiovanini ,
There are several things that looks strange to me in your sketch:
* there are 3 Chip Select defined for 2 SPI / 2 devices: 
`LTC_CS`, `sspin`, and `slaveSelectPin`
* I don't catch why you poll on MOSI signal ( `while(digitalRead(LTC2400_SDO){})` )
* I don't think that SPI_CONTINUE is the issue, it just mean that SPI library should not deselect device via SS pin, but in your case SS pin is not managed by SPI library it is manage by your application code (because you made the choice to no provide ss pin to` transfer() `function).
In other words, this parameter is useless in your case and you could removed it (as it is optional).
* I don't see SPI1 `beginTransaction()` 
Also I don't see `maxWrite() `call
* in `maxWrite()`, `address`, `value1 `and `value2 `parameters are declared as `int` (aka 32bit), and they are data to transfer,
but in SPI function `transfer()` data should be `uint8_t` (aka 8bit). If needed 16 bit data can be transfer with function `transfer16()` 
Make sure to pass right type to `transfer()`

**Comment by fpgiovanini** [comment]
Hi Alexandre.
Thanks for your repply.

there are 3 Chip Select defined for 2 SPI / 2 devices:
LTC_CS, sspin, and slaveSelectPin
A:LTC_CS-chip select for the MCP3551 voltage measurement in SPI2
sspin-chip select for the MCP3551 current measurement in SPI2
slaveSelectPin- chip select for the DAC on SPI1



I don't catch why you poll on MOSI signal ( while(digitalRead(LTC2400_SDO){}) )

A:A High state means the device is currently
performing an internal conversion and data cannot be
clocked out. A Low state means the device has finished
its conversion and the data is ready for retrieval on the
falling edge of SCK




I don't think that SPI_CONTINUE is the issue, it just mean that SPI library should not deselect device via SS pin, but in your case SS pin is not managed by SPI library it is manage by your application code (because you made the choice to no provide ss pin totransfer()function).
In other words, this parameter is useless in your case and you could removed it (as it is optional).
I don't see SPI1 beginTransaction()
A:
SPI.begin();////Initiallize the SPI 1 AS DEFAULT PARAMETERS
SPI_2.begin(); //Initiallize the SPI 2 
SPI_2.beginTransaction(SPI_SETTINGS);



Also I don't see maxWrite() call
A:Because I send you just a minimal skecth 

in maxWrite(), address, value1 and value2 parameters are declared as int (aka 32bit), and they are data to transfer,
but in SPI function transfer() data should be uint8_t (aka 8bit). If needed 16 bit data can be transfer with function transfer16()
Make sure to pass right type to transfer()
A:OK, you are right.I just need tranfers 16 bits in two bytes

**Comment by ABOSTM** [comment]
OK so as I understand there are 2 device on SPI2 and 1 on SPI1.
As a minimal sketch I would prefer to have only 1 device handled in SPI2 and 1 on SPI1 .
And as I said, minimal sketch `which reproduce the issue`  so a complete sketch which compiles and that you tested. Thus we avoid any other issues link to other part of the application. Start first with a simple sketch with a single transfert (not in loop) on SPI2 (1 ADC read) which should work, then add a single transfer on SPI1 (DAC output if I understand correctly)  if it fails you have the minimal sketch, if it works you can add step by step other SPI transfers, or do transfer in loop, ...  and you will have good hint on what fails.
